addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js

```diff
@@ -2766,12 +2766,68 @@ export class OdooEditor extends EventTarget {
         if (this._currentMouseState === 'mouseup') {
             this._fixFontAwesomeSelection();
         }
-        if (
-            selection.rangeCount &&
-            selection.getRangeAt(0) &&
-            this.options.onCollaborativeSelectionChange
-        ) {
-            this.options.onCollaborativeSelectionChange(this.getCurrentCollaborativeSelection());
+        if (selection.rangeCount && selection.getRangeAt(0)) {
+            // Handle selection/navigation at the edges of links.
+            const link = getInSelection(this.document, 'a');
+            if (link) {
+                // 1. If the selection starts or ends at the beginning of a link
+                //    (before the start zws (SZWS)), move the selection after
+                //    the SZWS. This will prevent the selection from being
+                //    ejected from the link when typing (because the SZWS is not
+                //    editable).
+                const startZws = link.querySelector('[data-o-link-zws="start"]');
+                const linkLeader = startZws && !selection.isCollapsed && ['anchor', 'focus'].find(point => (
+                    // (in link at offset 0 with no previous leaf in link)
+                    !selection[`${point}Offset`] && closestElement(selection[`${point}Node`], 'a') === link &&
+                    closestElement(previousLeaf(selection[`${point}Node`], this.editable), 'a') !== link
+                ));
+                if (linkLeader) {
+                    const [node, offset] = [startZws.nextSibling || link, startZws.nextSibling ? 0 : 1];
+                    if (linkLeader === 'anchor') {
```

Comment:
Yes it was to avoid duplicating the check because this is in `onselectionchange` so I'd want it to be as fast as possible, and because that check is already not so simple, I'd prefer it if I could not have to read it twice ^^'

Reply:
I think it's more confusing the way it is now. We could try having both version side by side to check.

Reply:
I'll write down the two versions here so you can compare.

Reply:
The version with duplication:
```js
const startZws = link.querySelector('[data-o-link-zws="start"]');
if (
    startZws &&
    !selection.isCollapsed &&
    // The anchor is at the beginning of the link, ie. at offset
    // 0 of the link, with no previous leaf that is in the link.
    !selection.anchorOffset &&
    closestElement(selection.anchorNode, 'a') === link &&
    closestElement(previousLeaf(selection.anchorNode, this.editable), 'a') !== link
) {
    setSelection(startZws.nextSibling || link, startZws.nextSibling ? 0 : 1, selection.focusNode, selection.focusOffset);
    return; // The selection is changed and will therefore re-trigger the _onSelectionChange.
} else if (
    startZws &&
    !selection.isCollapsed &&
    // The focus is at the beginning of the link, ie. at offset
    // 0 of the link, with no previous leaf that is in the link.
    !selection.focusOffset &&
    closestElement(selection.focusNode, 'a') === link &&
    closestElement(previousLeaf(selection.focusNode, this.editable), 'a') !== link
) {
    setSelection(selection.anchorNode, selection.anchorOffset, startZws.nextSibling || link, startZws.nextSibling ? 0 : 1);
    return; // The selection is changed and will therefore re-trigger the _onSelectionChange.
}
```
vs without:
```js
const startZws = link.querySelector('[data-o-link-zws="start"]');
const linkLeader = startZws && !selection.isCollapsed && ['anchor', 'focus'].find(point => (
    // The point is at the beginning of the link, ie. at offset
    // 0 of the link, with no previous leaf that is in the link.
    !selection[`${point}Offset`] && closestElement(selection[`${point}Node`], 'a') === link &&
    closestElement(previousLeaf(selection[`${point}Node`], this.editable), 'a') !== link
));
if (linkLeader) {
    const [node, offset] = [startZws.nextSibling || link, startZws.nextSibling ? 0 : 1];
    if (linkLeader === 'anchor') {
        setSelection(node, offset, selection.focusNode, selection.focusOffset);
    } else {
        setSelection(selection.anchorNode, selection.anchorOffset, node, offset);
    }
    return; // The selection is changed and will therefore re-trigger the _onSelectionChange.
}
```

Reply:
Conclusion: keep version without duplication but with `[[anchorNode, anchorOffset], [focusNode, focusOffset]]` instead of `['anchor', 'offset']` and check `linkLeader[0] === anchorNode` instead of `linkLeader === 'anchor'`.